### PR TITLE
[llvm][dfa-jump-threading] Allow DFAJumpThreading with optsize

### PR DESCRIPTION
**Pull request by ilovepi**
Though DFAJumpThreadingPass typically introduces size increases, targets
optimizing for size may want to enable the pass for certain TUs and
workloads. This change adds options to override the default behavior,
and enable the DFAJumpThreadingPass to run, even when compiling for
size.

To experiment with DFAJumpThreading and with `minsize`,
pass `-enable-dfa-jump-optsize` to the backend.

### CODE REVIEW

File: llvm/lib/Passes/PassBuilderPipelines.cpp [hunk 1]
```diff
@@ -222,6 +222,11 @@ static cl::opt<bool>
     EnableDFAJumpThreading("enable-dfa-jump-thread",
                            cl::desc("Enable DFA jump threading"),
                            cl::init(false), cl::Hidden);
+static cl::opt<bool>
+    DFAJumpThreadingOptSize("dfa-jump-thread-optsize",
+                           cl::desc("Enable DFA jump threading when optimizing for size"),
+                           cl::init(false), cl::Hidden);
+
```
Comment (ilovepi) [inline llvm/lib/Passes/PassBuilderPipelines.cpp:229]: Its rather unfortunate that we need 2 flags to enable a single behavior ... Is there a common place we can put this flag and use it here and in `DFAJumpThreading.cpp`?

**Comment by ilovepi** [comment]
cc: @petrhosek @PiJoules since I think they're interested.

**Comment by aeubanks** [comment]
Can we have a more principled approach to determine when to allow this pass to run than a `cl::opt`? For example, change the `CostThreshold` in the pass depending on whether or not the function has `optsize`.

**Comment by ilovepi** [comment]
> Can we have a more principled approach to determine when to allow this pass to run than a `cl::opt`? For example, change the `CostThreshold` in the pass depending on whether or not the function has `optsize`.

Sure. I can take a look at doing it that way instead. it will probably be much cleaner.

**Comment by nikic** [comment]
This pass is not part of the default pipeline -- do you enable this pass downstream, or do you want to add options for *all* uses of minsize, even if there is no evidence of usefulness?

**Comment by ilovepi** [comment]
> This pass is not part of the default pipeline -- do you enable this pass downstream, or do you want to add options for _all_ uses of minsize, even if there is no evidence of usefulness?

We’d like to see if this can be useful to some of our size constrained users in the embedded space. Particularly because it has been proposed to turn this on in the default pipeline.  https://discourse.llvm.org/t/rfc-enable-dfa-jumpthreading-pass-by-default/77231. Currently we don’t have evidence that this is beneficial, but for these targets we also don’t have a way to vet that with the current implementation.  

So to be concrete, I’d like the ability to enable this pass somehow even if compiling for size. If we can control this behavior with a threshold, then mechanically I don’t see much difference between that an the flags I’ve added in this pass.  But right now we don’t have a way to evaluate if this may be useful at any threshold, and I think it’s useful if we can turn this knob in some way without changing the default behavior.

**Comment by aeubanks** [comment]
also, we should remove pipeline checks for "optsize" and move those checks into the passes themselves

**Comment by ilovepi** [comment]
> also, we should remove pipeline checks for "optsize" and move those checks into the passes themselves

done in https://github.com/llvm/llvm-project/pull/83188

**Comment by ilovepi** [comment]
> Can we have a more principled approach to determine when to allow this pass to run than a `cl::opt`? For example, change the `CostThreshold` in the pass depending on whether or not the function has `optsize`.

So, I've experimented with this, and I'm not sure this is the right approach to take right away. I think its premature to suppose that at `-Oz` we would ever want this pass on by default. I don't think its a good idea to try and guess what the threshold should be. Likely, that is something specific targets would need to tune on their own via `-dfa-cost-threshold=`. I don't see a better way to allow us to experiment with this, than adding this flag, which is simple(allow the pass w/ optsize), and we can still use `-dfa-cost-threshold=` to determine what we should choose.

I'd like to avoid running afoul of this pass being on by default in the future for size optimized targets when we're still in the experimental stage of evaluating this passes usefulness when combined with size optimizations. I agree that if we believe this is a good default in the future, we should adjust the logic to modify the threshold for optsize, but right now I don't have confidence that is the correct way forward.

I'm looking to see if there is a way to use -enable-dfa-jump-threading to control this behavior, instead of introducing a new cl::opt.

**Comment by ilovepi** [comment]
So, if DFAJumpThreading is ever on by default, then I think its pretty simple to repurpose `-enable-dfa-jump-thread` to enable the behavior in the pass, and we can drop `-enable-dfa-jump-optsize`  @aeubanks, does that sound like a better plan?

**Comment by aeubanks** [comment]
> I don't think its a good idea to try and guess what the threshold should be. Likely, that is something specific targets would need to tune on their own via -dfa-cost-threshold=.
The inliner has a target-independent threshold and it's much more important than this pass, so I don't see why this pass can't do the same thing.

As for experimenting, I was hoping that you could just experiment locally with local changes to LLVM. Is it much easier to experiment with a flag in LLVM release? If so then this is fine.

This pass is already experimental (given it's not enabled by default) so I'm not too worried about regressing it in terms of binary size.

**Comment by ilovepi** [comment]
Well, since I'm not going to have time to work on this in the near future, and it will be quite a while before I can roll out an experimental toolchain to partners, I'm going to close this for now. Its easy enough to re-open when we have  better feeling for its value, but I do still think its good to have a way to opt into a transform irrespective of the optlevel.

**Pull request closed without merging**